### DISCUSSION TITLE
Extend support for batch operation mocking / assertions

### DISCUSSION THREAD

**Discussion by acline443** [pashagolub/pgxmock#152] [Q&A]
### Details

I may be misusing the library, so correct me if I am wrong. But I would like to write unit tests for functions that queue batches and then send the batch at the end. I would like to assert on the Updates/Insertions that are executed. 

It seems to me like this is not currently supported as the SendBatch method does basically nothing:
```go
func (c *pgxmock) SendBatch(context.Context, *pgx.Batch) pgx.BatchResults {
	return nil
}
```

Best assertion I can do is assert that the queued batch length is greater than 0. 

<details>
<summary>Checklist</summary>

- [X] `pgxmock.go`
> * Add a new type `ExpectedBatch` with fields for the expected SQL statements and their arguments, and methods to set the expected results or errors.
> • Modify the `SendBatch` method to return mock batch results based on the queued expectations.
> • Add a new method `ExpectBatch` to the `pgxmock` type to queue an expected batch operation. This method should return an `ExpectedBatch` instance.
> • Update the `ExpectationsWereMet` method to also check if all expected batch operations were met.

- [ ] `pgxmock_test.go`
> * Add unit tests for the new `ExpectBatch` method and the modified `SendBatch` method. These tests should cover various scenarios, including successful batch operations, batch operations that result in errors, and batch operations that do not meet the expectations.

</details>

**Comment by pashagolub** [comment]
Yes, unfortunately, `SendBatch()` is not currently supported. I do have plans for it. However, this will require some spare weeks to implement. One needs to implement [pgx.BatchResults](https://pkg.go.dev/github.com/jackc/pgx/v5#BatchResults) from scratch and incorporate this into the expectations paradigm. Of course, pull requests are more than welcome!

**Reply by makubit** [reply]
Hi :) 
I started working on this feature on my fork: https://github.com/makubit/pgxmock/tree/feature/152-support-batch-operations
It doesn't have all the tests yet, but most likely I will finish it next week. You can check if this implementation is suitable, I can create a PR once everything is completed.

There are some new builder functions to help define the batch structure as well as the BatchResults interface.
I can create documentation for this part as well, as it's something new.

Hope you'd like it (right now I'm using `.Exec()` and `.Close()` in my private project, so I still have to test `.Query()` and `.QueryRow()` from BatchResults interface).

**Reply by pashagolub** [reply]
Wow! Thanks for working on this. I had a quick look over and it looks nice. Did you consider to use [reflect.DeepEqual()](https://pkg.go.dev/reflect#DeepEqual) instead of writing your own comparators? My main concern is introduction of `unsafe` in source code. Usually it means something was overengineered.

I'll try to spend some more time on this after [pgconf.eu](https://2023.pgconf.eu/).

Thanks for your time and efforts!

**Reply by makubit** [reply]
The issue is that `pgx.Batch` type supports only two functions - `Len()` and `Queue()` and has inside unexported slice of already queued queries. I didn't know how to better reflect this `Value`, as `DeepEqual()` didn't work in this case. I will think about your suggestion, maybe something will come up to my mind.

Thanks for looking at this code ;)

**Reply by pashagolub** [reply]
Yeap, that's unfortunate. I thought `reflect.DeepEqual` is able to compare structs:

>...
Struct values are deeply equal if their corresponding fields, both exported and unexported, are deeply equal.

**Reply by makubit** [reply]
The struggle with that is - that you have to batch the exact queries, I wanted a similar solution, that is used for `Query`, so in batch, we can reuse `queryMatcher` and also `argsMatcher` functionality. It, of course, doesn't need to be done this way, but `DeepEqual` has some limitations.
I'll think about your suggestion :)

**Reply by pashagolub** [reply]
I see now. That's not a simple question indeed :)
